### PR TITLE
memtest86-efi: init at 8.0

### PR DESCRIPTION
**Pull request by cdepillabout**
This updates memtest86 to (almost) the latest version.

memtest86-4.3.6 is the older version.  It was open source, but marked broken.  memtest86-8.0 is the newer version.  It is no longer open source, but it is possible to build and use.  There appear to be no restrictions on its use or redistribution.

###### Motivation for this change

The old memtest86 was marked as broken.

I plan to send a future PR that actually adds memtest86 as an option for the bootloader.
###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by cdepillabout** [comment]
I wonder if ofborg is able to compile non-free derivations?

@GrahamcOfBorg build memtest86

**Comment by JohnAZoidberg** [comment]
Hmmm, I propose to keep the free version as the default.
There has been a fork, although not quite as well maintained: https://www.memtest.org

File: pkgs/tools/misc/memtest86/default.nix [hunk 1]
```diff
@@ -1,32 +1,51 @@
...
+    # the ./BOOT/ directory with the memtest86 EFI app.
+    guestfish --ro --add ./memtest86-usb.img --mount /dev/sda1:/  copy-out /EFI/BOOT .
+
+    cp -r BOOT/* $out/
   '';
 
-  meta = {
+  meta = with lib; {
     homepage = http://memtest86.com/;
+    downloadPage = https://www.memtest86.com/download.htm;
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/memtest86/default.nix]: Reminder: RFC45 is likely to be accepted soon. Meaning we will deprecate unquoted URL syntax.

```suggestion
    downloadPage = "https://www.memtest86.com/download.htm";
```

File: pkgs/tools/misc/memtest86/default.nix [hunk 2]
```diff
@@ -1,32 +1,51 @@
...
     description = "A tool to detect memory errors, to be run from a bootloader";
-    broken = true;
+    # The Memtest86 License for the Free Edition states,
+    # "MemTest86 Free Edition is free to download with no restrictions on usage".
+    # However the source code for Memtest86 does not appear to be available.
+    #
+    # https://www.memtest86.com/license.htm
+    license = licenses.unfreeRedistributable;
```
Comment (c0bw3b) [inline pkgs/tools/misc/memtest86/default.nix]: At first glance it's not even clear redistribution is allowed

**Review by c0bw3b — CHANGES_REQUESTED** [review]
> MemTest86 V8 images support only UEFI boot. On machines that don't support UEFI, MemTest86 will not boot. Please download the older V4 BIOS release of MemTest86 instead.

This should be its own derivation (`memtest86-efi` for example) and we should keep the current `memtest86` as-is for non-UEFI systems.

File: pkgs/tools/misc/memtest86/default.nix (hunk 1 above)
Comment (cdepillabout) [reply pkgs/tools/misc/memtest86/default.nix]: Thanks!

**Comment by cdepillabout** [comment]
@c0bw3b @JohnAZoidberg Thanks for taking the time to review this.  You both had similar suggestions, so let me respond to them together:

> I propose to keep the free version as the default.

> This should be its own derivation (memtest86-efi for example) and we should keep the current memtest86 as-is for non-UEFI systems.

This sounds like a good plan.  I'll make this change.  I'll call this package `memtest86-efi` as suggested.

The reason I didn't do this from the beginning is because the current `memtest86` derivation is (extremely) out-of-date and marked as broken.  The company creating MemTest86 no longer supports it.  It is questionable as to whether or not anyone would ever want to use it.  I believe that the `memtest86` from this PR is the one that will be more heavily used going forward.

**Comment by cdepillabout** [comment]
@JohnAZoidberg 

>There has been a GPL fork, although not quite as well maintained: https://www.memtest.org

It looks like this is in the nixpkgs repo as `memtest86plus`, but it doesn't support EFI :-\

File: pkgs/tools/misc/memtest86/default.nix (hunk 2 above)
Comment (cdepillabout) [reply pkgs/tools/misc/memtest86/default.nix]: The license page on the memtest86.com site is completely lacking any sort of helpful information :-\

The redistribution story is clarified a little in a response on their forum:

https://www.passmark.com/forum/memtest86/44494-version-8-1-distribution-file-is-not-versioned?p=44505#post44505

**Comment by c0bw3b** [comment]
Thanks for explaining. :)

Regarding base memtest86 age -> I'd rather alias the current `memtest86` to `memtest86plus` which is still free/libre and [receives updates from coreboot](https://review.coreboot.org/cgit/memtest86plus.git/).
(I'm willing to craft a PR for this)

Since this new variant is unfree users will need additional steps to use it. That's why I don't think it should take over the current `memtest86` attribute name.

**Comment by cdepillabout** [comment]
@c0bw3b 

> Since this new variant is unfree users will need additional steps to use it. That's why I don't think it should take over the current memtest86 attribute name.

Normally I would agree with you, but in this specific case `memtest86` has been broken since at least 16.09, so it is unlikely there is anyone still using it.  It is unlikely anyone will be surprised when it starts to require `allowUnfree`.

Also, I don't think it makes sense to have three derivations `memtest86`, `memtest86plus` and `memtest86-efi` when there are really only two pieces of software: MemTest86 and MemTest86+.  I also don't think that `memtest86` should be an alias for `memtest86plus`, since they are really two different pieces of software.

That being said, I don't actually care about the name of the derivation that much, so if you're saying that you'll only merge in this PR if it is changed to be `memtest86-efi`, then I'll gladly make that change!

**Comment by cdepillabout** [comment]
Also, in case anyone is interested, here is a nixos module for installing memtest86 with systemd-boot:

https://github.com/cdepillabout/nixpkgs/commit/8c85e24aa65c69d647637f02b01eb6f97d11378c

I plan on sending this module as a PR as soon as this PR is merged in.

**Comment by c0bw3b** [comment]
> Normally I would agree with you, but in this specific case `memtest86` has been broken since at least 16.09, so it is unlikely there is anyone still using it.

Actually on NixOS people would use it through the config option `boot.loader.grub.memtest86.enable` which use `memtest86plus` already. Probably explains why the broken status of legacy memtest86 remained unnoticed until now. In effect we were already using memtest86plus as the continuation of legacy memtest86 free.

> It is unlikely anyone will be surprised when it starts to require `allowUnfree`.

I'm not so sure but the important point here is that allowUnfree has legal consequences. On a server in a corporate context it's likely forbidden. Somehow I find mapping a proprietary software to a formerly free attribute name _might_ encourage _some_ users to mindlessly switch that boolean to true just to be able to go on. And I find myself uneasy with that. This is just my own subjective opinion and I'd welcome other contributors' input.

> when there are really only two pieces of software: MemTest86 and MemTest86+

But that is not the case : MemTest86 V4 and MemTest86 V8 are _not_ the same software at all. One is free and BIOS-only, the other is proprietary and UEFI-only. The confusion stems from PassMark decisions.

> That being said, I don't actually care about the name of the derivation that much, so if you're saying that you'll only merge in this PR if it is changed to be `memtest86-efi`, then I'll gladly make that change!

I actually looked at Arch/AUR and the name comes from them. I genuinely think it clarifies the difference between V4 (deprecated) and V8.

**Comment by cdepillabout** [comment]
@c0bw3b Okay, that sounds good.

I just pushed commit 439a755 that renames this to `memtest86-efi`.

This should be ready for a final review and merge.

**Comment by c0bw3b** [comment]
@GrahamcOfBorg build memtest86-efi

**Comment by c0bw3b** [comment]
:man_facepalming: moment / of course ofBorg won't built it

**Review by c0bw3b — APPROVED** [review]

**Comment by Lassulus** [comment]
tested with nix-review.

**Pull request merged**